packages/vats/src/orchestration.js

```diff
@@ -0,0 +1,255 @@
+// @ts-check
+/** @file Orchestration service */
+import { makeTracer } from '@agoric/internal';
+import { V as E } from '@agoric/vat-data/vow.js';
+import { M, matches } from '@endo/patterns';
+import { AmountShape, BrandShape } from '@agoric/ertp';
+import '@agoric/network/exported.js';
+
+const { Fail, bare } = assert;
+const trace = makeTracer('Orchestration');
+
+/** @typedef {string} ChainAddress */
+
+/**
+ * @typedef {object} OrchestrationPowers
+ * @property {ERef<
+ *   import('@agoric/orchestration/src/types').AttenuatedNetwork
+ * >} network
+ */
+
+/**
+ * @typedef {MapStore<
+ *   keyof OrchestrationPowers,
+ *   OrchestrationPowers[keyof OrchestrationPowers]
+ * >} PowerStore
+ */
+
+/**
+ * @template {keyof OrchestrationPowers} K
+ * @param {PowerStore} powers
+ * @param {K} name
+ */
+const getPower = (powers, name) => {
+  powers.has(name) || Fail`need powers.${bare(name)} for this method`;
+  return /** @type {OrchestrationPowers[K]} */ (powers.get(name));
+};
+
+export const ChainAccountI = M.interface('ChainAccount', {
+  getAddress: M.call().returns(M.string()),
+  getLocalAddress: M.callWhen().returns(M.string()),
+  executeTx: M.callWhen(M.arrayOf(M.record())).returns(M.any()),
+  executeEncodedTx: M.callWhen(M.string()).returns(M.any()),
+  deposit: M.callWhen(M.remotable('Payment'))
+    .optional(M.pattern())
+    .returns(AmountShape),
+  getPurse: M.callWhen(BrandShape).returns(M.remotable('VirtualPurse')),
+  close: M.callWhen().returns(M.promise()),
+});
+
+/** @param {import('@agoric/base-zone').Zone} zone */
+const prepareChainAccount = zone =>
+  zone.exoClass(
+    'ChainAccount',
+    ChainAccountI,
+    /**
+     * @param {Connection} connection
+     * @param {Port} port
+     * @param {ChainAddress} remoteAddress
+     */
+    (connection, port, remoteAddress) => ({
+      connection,
+      port,
+      remoteAddress,
+    }),
+    {
```

Comment:
should we also add a getPort() method, so an advanced consumer can retrieve the port they bound?

Reply:
Added in [148ebf2](https://github.com/Agoric/agoric-sdk/pull/9114/commits/148ebf2249fab5d4bcf960aa194375f3bcc959f1)